### PR TITLE
Add sample on how to localize in VS Code

### PR DESCRIPTION
**Pull request by vazexqi**
### What does this PR do?

As far as I can tell, there is not much documentation on how extension authors can localize their VS Code extensions. VS Code itself (and extensions that are bundled with it, are properly localized) but there are not many examples of how to do it (short of reading the source code). There have been several questions on this: 

* Microsoft/vscode-extension-samples#20
* Microsoft/vscode-nls#11
* Microsoft/vscode-nls#5

This is a small extension that shows several things:

1. What you can localize (strings in package.json and strings in your .ts/.js files)
1. How to localize them (create package.nls.{language}.json and the various .json file)
1. How you can do so automatically using the `vscode-nls` and `vscode-nls-dev` packages.

### What is not in the scope of this PR?

This PR does not talk about Transifex and different file formats for localization. The `vscode-nls-dev` module supports it but I intentionally left it out since the implementation seems very specific to the way Microsoft is doing it. This PR makes it simple by assuming that you have the required structure in the i18n folder (you either did it by hand or imported it using some other means)

It also tries to demystify a lot of the magic that `vscode-nls` is doing by suggesting manual but simpler ways for accomplishing the same thing.

### Disclosure

Another reason I am doing this is also to learn how localization works in VS Code and also validate some of my assumptions on what is save to use (API stable). We are creating several extensions and localization is something that we have been asked to look into. Even if this PR is not accepted, comments on the assumptions in my code is appreciated so that we can understand what we should or should not rely on.

### CODE REVIEW

File: i18n-sample/README.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
+# README
+## This is the README for the "i18n-sample" 
+-------------------
+
+This folder contains a sample VS code extension that shows how to use the
+package.nls.json and the vscode-nls library for localization. For this sample,
+it shows two commands: Hello and Bye in English and Japanese.
+
+**Assumptions**
```
Comment (vazexqi) [inline i18n-sample/README.md]: Are these assumptions valid?

File: i18n-sample/README.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+{
+    "extension.sayHello.title": "Hello",
+}
+
+```
+
+Then, create the corresponding package.nls.{your_language}.json files for each language to localize.
+
+2. It is also possible to use your own library for localizing text in your
+   source file. You would use the value of `process.env.VSCODE_NLS_CONFIG`
```
Comment (vazexqi) [inline i18n-sample/README.md]: Is this environment variable part of the API?

File: i18n-sample/gulpfile.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  return function() {
+    const input = es.through();
+    const tsFilter = filter(['**/*.ts', '!**/vscode-nls-dev/lib/**'], {
+      restore: true
+    });
+
+    const output = input
+      .pipe(tsFilter)
+      .pipe(compilation())
+      .pipe(build ? nlsDev.rewriteLocalizeCalls() : es.through())
```
Comment (vazexqi) [inline i18n-sample/gulpfile.js]: Is using`vscode-nls-dev` always going to require the use of gulp in the build pipeline?

**Comment by dbaeumer** [comment]
@vazexqi none of the files do have a copyright statement. Do you want to add some or can I add the standard MS one.

**Comment by dbaeumer** [comment]
@vazexqi I worked on the PR and made a couple of changes, mainly around structure and things that have changed in our thinking around nls in the meantime.

The demo.gif got outdated through that. Could you please generate a new one. The change I made are available here: https://github.com/Microsoft/vscode-extension-samples/tree/vazexqi-master

**Comment by vazexqi** [comment]
@dbaeumer - Thanks for the review and changes.

1. I'll add the standard Microsoft one of to all the .ts. files
```
/*---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------*/
```
2. I'll update the gif

I should be able to do this over the weekend and have an updated PR in by Monday.

**Comment by vazexqi** [comment]
@dbaeumer I've rebased master of https://github.com/vazexqi/vscode-extension-samples onto the latest master of https://github.com/Microsoft/vscode-extension-samples and cherry-picked your changes over since I don't have access to push to vazexqi-master on this repo.

I've also regenerated the gif so that it shows using `gulp package`.

I noticed that you removed two files (i18n-sample/i18n/jpn/out/command/sayBye.i18n.json and i18n-sample/i18n/jpn/out/extension.i18n.json). This might have been because your .gitignore is too strict. If you don't have those files, then when you run `gulp package` you are going to see

```
[17:17:04] [i18n] Generating localized messages for 'jpn' resulted in the following problems:
[17:17:04] [i18n]
[17:17:04] [i18n] Message file jpn/out/extension.i18n.json not found. Missing messages: 1
[17:17:04] [i18n]
[17:17:04] [i18n]
[17:17:04] [i18n] Generating localized messages for 'jpn' resulted in the following problems:
[17:17:04] [i18n]
[17:17:04] [i18n] Message file jpn/out/command/sayBye.i18n.json not found. Missing messages: 1
[17:17:04] [i18n]
[17:17:04] [i18n]
[17:17:04] Finished 'internal-nls-compile' after 799 ms
[17:17:04] Starting 'add-i18n'...
[17:17:04] [i18n] Generating localized messages for 'jpn' resulted in the following problems:
[17:17:04] [i18n]
[17:17:04] [i18n] Message file jpn/package.i18n.json not found. Missing messages: 2
[17:17:04] [i18n]
[17:17:04] [i18n]
```

I've added them back in.

**Comment by dbaeumer** [comment]
So, that was my mistake. I will tweak the git ignore and add the files back.

**Comment by dbaeumer** [comment]
Thanks again. Merge into master

**Pull request merged**